### ISSUE TITLE
cosine_similarity produces wrong output for tensors of different size

### ISSUE THREAD

**Issue by ramiro050** [pytorch/pytorch#109333]
### 🐛 Describe the bug

After https://github.com/pytorch/pytorch/pull/104771, the input tensors to `cosine_similarity` no longer get broadcasted to a common shape before taking their norm. This sometimes leads to `cosine_similarity` greater than 1 when tensors have different sizes.

```python
In [1]: a = torch.ones(2, 3, dtype=torch.float)

In [2]: b = torch.ones(1, 1, dtype=torch.float)

In [3]: torch.cosine_similarity(a, b)
Out[3]: tensor([1.7321, 1.7321])
```

Moreover, if one of the tensors has rank smaller than or equal to the `dim` argument of `cosine_similarity`, an out-of-range error is generated, despite [the documentation](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html) saying that this should be a valid case: 

> x1 and x2 must be broadcastable to a common shape. dim refers to the dimension in this common shape.

```python
In [8]: a = torch.ones(2, 3, dtype=torch.float)

In [9]: b = torch.ones(1, dtype=torch.float)

In [10]: torch.cosine_similarity(a, b)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[10], line 1
----> 1 torch.cosine_similarity(a, b)

IndexError: Dimension out of range (expected to be in range of [-1, 0], but got 1)
```

### Versions

```
PyTorch version: 2.2.0.dev20230909+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux rodete (x86_64)
GCC version: (Debian 12.2.0-14) 12.2.0
Clang version: 14.0.6
CMake version: version 3.25.1
Libc version: glibc-2.37

Python version: 3.11.5 (main, Aug 25 2023, 07:43:52) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-6.3.11-1rodete2-amd64-x86_64-with-glibc2.37
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          96
On-line CPU(s) list:             0-95
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) CPU @ 2.00GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
Stepping:                        3
BogoMIPS:                        4000.32
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       1.5 MiB (48 instances)
L1i cache:                       1.5 MiB (48 instances)
L2 cache:                        48 MiB (48 instances)
L3 cache:                        77 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-23,48-71
NUMA node1 CPU(s):               24-47,72-95
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX flush not necessary, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Mitigation; IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; Clear CPU buffers; SMT Host state unknown

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] torch==2.2.0.dev20230909+cpu
[pip3] torchvision==0.17.0.dev20230909+cpu
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @kadeng @albanD

**Comment by ramiro050** [comment]
@lezcano @ngimel

**Comment by lezcano** [comment]
oops. I'll submit a fix

**Comment by gdippolito** [comment]
Hello!
Would it be possible to include this bug fix in 2.1.2? 

I think this doesn't work as intended, for people that are using Pytorch 2.1.x they would need to cherry-pick the commit and build from source? Thanks 🙏

**Comment by huydhn** [comment]
Let me double check and will post a cherry pick for this if it satisfies the requirements

**Comment by huydhn** [comment]
Confirm that this is fixed with 2.1.2.  The final RC of 2.1.2 can be installed from `pip install --pre torch==2.1.2 --index-url https://download.pytorch.org/whl/test`
